### DISCUSSION TITLE
RFC: Stop exporting symbols

### DISCUSSION THREAD

**Discussion by flavorjones** [sparklemotion/nokogiri#2746] [RFCs]
This document is a request for comments, specifically from downstream maintainers of gems that integrate with Nokogiri's C API, or the C API of libxml2, libxslt, or libgumbo.

## Proposal

Limit the symbols that are exported by the Nokogiri shared library (extension) to the bare minimum necessary for Nokogiri to operate. This might be the single symbol `Init_nokogiri`, but I need to do some exploration to be sure.

## Context

Nokogiri has always _exported_ the symbols from libxml2 and libxslt that are statically linked into the extensions. On Linux this has led to one or two issues with symbol collisions (see #2106), but with careful naming we've mostly avoided problems.

In the final days of shipping Nokogiri v1.14.0 with native (precompiled) support for Ruby 3.2, we're struggling a bit with symbol resolution. Ruby 3.2, when compiling on Darwin, now uses the `-bundle_loader` linker flag to resolve symbols against the Ruby executable as if it were a shared library. This means that, when running a Ruby compiled with the `--enable-shared` flag, that the extension will fail to resolve Ruby symbols like `rb_cObject`. We can work around that with the `-flat_namespace` linker flag, which mimics the behavior we already see on Linux and allows us to resolve these symbols at runtime. But for reasons I don't fully understand, many Rubies on Darwin seem to load the libxml2 and libxslt dylibs that ship with XCode commandline tools ("CLT"), and so _every_ libxml2 symbol is a collision and resolves to the _wrong_ libxml2 (not the version we've patched and statically linked into the extension).

To work around this last problem, the best solution we know of right now seems to be to avoid exporting those symbols by using the `-load_hidden` flag (or a similar mechanism, there are several we could choose from).

The decision I think we've made is to avoid exporting these symbols _only_ in the precompiled Ruby 3.2 Darwin extensions. But that opens up the question of: should we avoid exporting these symbols everywhere, both for consistency and for reliability. _That_ is the question asked by this RFC.

## Impact

**Good**: This would prevent accidental symbol collisions such as the https://github.com/sparklemotion/nokogiri/pull/2106 on Linux, and would ensure that we always pull in the desired version of libxml2, avoiding problems like the ones we're currently experiencing with Ruby 3.2 (see https://github.com/rake-compiler/rake-compiler-dock/issues/87 for extended discussion and more links).

**Bad**: This would also, however, prevent a small but non-zero number of downstream gems from integrating with Nokogiri's C API, or the C API of libxml2, libxslt, or libgumbo. A notable gem that did this was https://github.com/rubys/nokogumbo (now merged into Nokogiri itself). Another notable gem that I know that does this is `nokogiri-xmlsec` (and the various forks of it, the most popular seems to be https://github.com/instructure/nokogiri-xmlsec-instructure). So this may prevent experimentation and innovation (see Nokogumbo) as well as putting hurdles in front of useful integrations like xmlsec.

## Feedback

I would like to hear from downstream gem maintainers about how they're using the C APIs exported by Nokogiri's extension today. Please comment below!

## References

- [nokogiri/2022-12-darwin-symbol-resolution.md at main · sparklemotion/nokogiri · GitHub](https://github.com/sparklemotion/nokogiri/blob/main/adr/2022-12-darwin-symbol-resolution.md)
- https://github.com/ruby/ruby/commit/50d81bf
- [[Ruby 3.2] having runtime issues on darwin · Issue #87 · rake-compiler/rake-compiler-dock](https://github.com/rake-compiler/rake-compiler-dock/issues/87)
- [dep: add ruby 3.2 support by flavorjones · Pull Request #2732 · sparklemotion/nokogiri](https://github.com/sparklemotion/nokogiri/pull/2732)
- https://github.com/stevecheckoway/bundle_test and [explore whether \`-load\_hidden\` will work around flat namespace by flavorjones · Pull Request #1 · stevecheckoway/bundle\_test](https://github.com/stevecheckoway/bundle_test/pull/1)
- [Video from Apple explaining the Darwin toolchain changes](https://developer.apple.com/videos/play/wwdc2022/110362/)

**Comment by flavorjones** [comment]
I'm trying out this approach with the `sqlite3-gem` in this draft PR: https://github.com/sparklemotion/sqlite3-ruby/pull/371

**Reply by flavorjones** [reply]
Note for posterity: it worked, and this is now how sqlite3-ruby ships.

**Comment by maths22** [comment]
https://github.com/instructure/nokogiri-xmlsec-instructure currently uses `noko_xml_document_pin_node` and `noko_xml_node_wrap`, which if i'm understanding this proposal correctly would go away under the proposal.  It does rely on libxml symbols as well, but I believe gets those from the linker loading libxmlsec1's dependencies, so it doesn't depend on nokogiri for them.  (I am neither the original author of that extension nor very experienced at ruby C extension development, so I am definitely far from an expert on how that gem works or if it is doing things optimally.  I did however inherit maintenance of it, so I do care about trying to prevent it from breaking.)

**Reply by flavorjones** [reply]
@maths22 Yeah, I'm curious what symbols you use. I can export some of the nokogiri symbols, but it's going to be challenging to continue to export libxml2 functions given where the Darwin toolchain is going. I expect it calls quite a few libxml2 functions either directly or indirectly (it looks like at least `xmlUnlinkNode` and `xmlFreeNode` are called).

Given nokogiri-xmlsec-instructure currently doesn't support compaction (see https://github.com/instructure/nokogiri-xmlsec-instructure/pull/17), I'm curious whether you'd be open to talk about integrating some (or all?) of this functionality into Nokogiri, to solve both problems? Or perhaps there's another architectural workaround that I haven't thought of. I'd be interested in talking to you about what we should do here!

**Reply by maths22** [reply]
I would definitely be interested in integrated the functionality into nokogiri rather than maintaining a whole separate thing that plugs in rather jankily.  I did verify that nokogiri 1.14.0.rc1 on my m1 mac running macos 13.1 passes the build, so presumably somehow we are getting the libxml symbols from somewhere else, but who knows where...  Interestingly we use that gem heavily in production (through https://github.com/instructure/ruby-saml2 ), and we don't get segfaults regularly in production, so presumably the buggy part of that library is something relatively narrow.

**Reply by flavorjones** [reply]
@maths22 I believe that you're running the RC successfully ... though please note that the `load_hidden` hack is only turned on with Ruby 3.2. In that case, it's likely that you're pulling in libxml2 symbols from Ruby itself (which, if it's built with XCode CLT, will (for reasons that still escape my understanding) have the XCode version of libxml2 loaded into memory), which is _probably_ not going to be a problem, but, like, there's no good way to guarantee success.

We pulled in nokogumbo in early 2021, and it wasn't hard to do. We might have to do a bit of work around compaction support, but that shouldn't be hard. My big question is whether you'd be interested in helping support that functionality? I'm not familiar with it at all, and I'm not a user of it. I hope to really dig in and get more familiar with the codebase in the near future.

**Reply by flavorjones** [reply]
For posterity, here's what the Nokogumbo merger planning looked like: https://github.com/sparklemotion/nokogiri/issues/2204

**Reply by flavorjones** [reply]
One more note, just to clarify how I'm thinking about making these decisions. I have an _assumption_ that there are very few gems that are linking against Nokogiri's libxml2 or libxslt symbols (C functions).

It feels likely that this number may in fact be 1 (`nokogiri-xmlsec`) in which case, that's a compelling reason to consider merging it into Nokogiri. This number may be higher -- and that would be surprising -- and in that case I would want to try harder to come up with a solution to meet all those gems' needs.

But in 14 years of maintaining Nokogiri, I only remember seeing two gems that linked against Nokogiri's libxml2: `nokogumbo` and `nokogiri-xmlsec`. If that holds true, that is a good indicator that we don't need to support exporting symbols.

**Reply by maths22** [reply]
Running this search `path:ext/**/*.c noko_  -path:*/nokogiri/*` in github codesearch returns only nokogiri-xmlsec-instructure and forks, which I would say is at least a good argument in favor of your hypothesis.  And I think it would be a lot stranger if we had a gem that linked against nokogiri's exported libxml symbols without linking against any of against nokogiri's own symbols, but people do surprising things sometimes.

**Reply by maths22** [reply]
Oh I missed the first message of your three.  I forgot to mention that my local test was with ruby 3.2.  I'd definitely be happy to help support the libxmlsec functionality in nokogiri.  I know what my company uses it for and why we use it, but I'd also need to do some digging to really properly understand *how* it does what it does; so far mostly I just touch whatever breaks on a ruby or nokogiri upgrade and move on with my life, so I can't claim to understand it *well*.  But given that we rely on its functionality for hundreds of thousands of saml logins a day it would probably be worthwhile getting to know it better :)

**Reply by stevecheckoway** [reply]
@maths22 there’s a decent chance that Nokogiri’s vendored libxml2 and whichever libxml2 is getting linked to Ruby (and that nokogiri-xmlsec is linking to) are in fact different versions of libxml2. They could be incompatible versions. I wouldn’t be shocked if there are subtle bugs as a result, particularly around line numbers. (Libxml2’s design assumes at most 65534 lines but has a compile-time hack to kinda support larger line numbers in some cases.)

I spent a while trying to figure out how to deal with this on macOS by getting the runtime linker, dyld, to do the right thing. I didn’t really succeed.

One option is using dlopen/dlsym to look up specific symbols. (I investigated a hand-rolled GOT/PLT-like approach—essentially caching the symbol per call-site using a macro. It worked but the programming experience wasn’t great.)

One thing I’d like to understand is all of the configurations we want to support and then see how we can make that work on all supported platforms. There are many tricky restrictions around (re-)exporting symbols, two-level namespaces, vendored vs. system libraries, bundles vs. dynamic libraries, and `—enable-shared` vs. `—disable-shared` Ruby versions. Trying to support everything is probably too painful. And that’s just macOS. ELF symbol resolution is different. I need to do a bunch of reading about symbol resolution on Windows to know how that differs from the others.

**Comment by flavorjones** [comment]
OK, I think we've made a decision, which is:

- we'll plan to __not export__ libxml2/libxslt symbols in an upcoming minor release
- I'd like to integrate the `nokogiri-xmlsec` features into Nokogiri before (or in) that release

Obviously we can still learn as we do this work, and may possible change direction, but I'd like to get started.

**Comment by flavorjones** [comment]
Fun and interesting twist: https://github.com/sparklemotion/nokogiri/issues/2898 shows that problems happen _to other gems_ when they accidentally pull in Nokogiri's libxml2.

This increases my confidence that we've chosen the right path in deciding to not export symbols in an upcoming minor release. Ugh, what a mess.

**Comment by flavorjones** [comment]
Also note that the YARP project just ran into this issue, and I proposed solving it by hiding symbols in https://github.com/ruby/yarp/pull/1069

**Reply by flavorjones** [reply]
And here's a followup commit I made to YARP introducing symbol visibility tests: https://github.com/ruby/yarp/pull/1110/commits/c87d91aac13592fd6777904f0f5305d3d4c5d8af

We will want to do something similar for Nokogiri

**Comment by flavorjones** [comment]
Interesting update here which may force a decision: [this Rails PR](https://github.com/rails/rails/pull/55303) introduces `libvips` into Rails, and if it loads before Nokogiri then it forces use of the system libxml2, which breaks things. See [this thread]((https://github.com/rails/rails/pull/55596) for repro details.

I probably need to make Nokogiri more robust in situations where it's dynamically linking against a different library than it was compiled against; but that's non-trivial. And making all the symbols private is a sure fix.

Tentatively, let's shoot to stop exporting symbols in v1.19. I'll have to spend some time with https://github.com/sparklemotion/nokogiri/issues/2902 to see if we can land everything.
